### PR TITLE
SR-8 Add question endpoints

### PR DESCRIPTION
**Pull request by Mario-659**

### CODE REVIEW

**Comment by Mendrejk** [comment]
First, quick observations:
- Why is the RestController called QuestionResource and not QuestionController?
- You should not create the TeamRepository with this pull request. I know that you need it and we've stumbled into this issue with @juleczkatruskaweczka as well. There is a task on Jira to add the TeamRepository. It needs to be completed and merged before this (any endpoint related) PR.

I'll try to do a more in-depth review later today ;).

**Comment by Mario-659** [comment]
It's the convention at my company but I will rename it. 

Check out my tests. They were quite challenging 😅

File: backend/recruitment-system/src/test/groovy/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/QuestionServiceSpec.groovy [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+        team1 = new Team(1, "team one", Set.of(), Set.of(), Set.of())
+        team2 = new Team(2, "team two", Set.of(), Set.of(), Set.of())
+
+        question1 = new Question(0, "sample content", 2, 3, 4, team1)
+        question2 = new Question(1, "sample content 2", 3, 4, 5, team1)
+        question3 = new Question(2, "sample content 3", 4, 5, 6, team2)
+    }
+
+    def "should return all questions from repository"() {
+        when:
```
Comment (Mendrejk) [inline backend/recruitment-system/src/test/groovy/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/QuestionServiceSpec.groovy:44]: add stories to all when and then blocks

**Review by Mendrejk — CHANGES_REQUESTED** [review]
Fix It!

File: backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/QuestionService.kt [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        )
+
+        return questionRepository.save(questionToUpdate)
+    }
+
+    fun deleteQuestion(questionId: Int, teamId: Int) {
+        val question = questionRepository.findById(questionId)
+        if (question.isPresent && question.get().team.id == teamId) {
+            questionRepository.delete(question.get())
+        }
```
[the remark this answers is not in the document]
Comment (Mario-659) [reply backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/QuestionService.kt]: Updated. Let me now if oneliners are fine or returns with curlies would be better

File: backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/QuestionService.kt [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    fun getAllQuestionsByTeamId(teamId: Int): List<Question> {
+        return questionRepository.findAllByTeamId(teamId).toList()
+    }
+
+    fun saveQuestion(question: Question, teamId: Int): Question {
+        validateQuestion(question)
+        validateTeamId(teamId)
+
+        val questionToSave = Question(
+            0,
```
[the remark this answers is not in the document]
Comment (Mario-659) [reply backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/QuestionService.kt:39]: What do you mean? Can I omit id with default constructor? I added default values in constructors because something wasn't working properly (maybe serializing, I'm not sure now but I will check this later)

File: backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/serializer/QuestionJsonSerializer.kt [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+            writeStringField("content", question.content)
+            question.minimumMinutes
+                ?. let {writeNumberField("minimumMinutes", it) }
+                ?: writeNullField("minimumMinutes")
+            writeNumberField("maximumMinutes", question.maximumMinutes)
+            writeNumberField("teamId", question.team.id)
+            writeEndObject()
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (Mario-659) [reply backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/serializer/QuestionJsonSerializer.kt:28]: Yeah, I tried to use adnotation but couldn't find the right one. This solution adds quite a lot of boilerplate code but on the other hand separates declaration of Entities from logic related to serializing

File: backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/controller/QuestionResource.kt [hunk 5]
```diff
@@ -0,0 +1,41 @@
+package pl.edu.pwr.samorzad.informatyzacja.recruitmentsystem.controller
+
+import org.slf4j.LoggerFactory
+import org.springframework.http.HttpStatus
+import org.springframework.http.ResponseEntity
+import org.springframework.web.bind.annotation.*
+import pl.edu.pwr.samorzad.informatyzacja.recruitmentsystem.model.Question
+import pl.edu.pwr.samorzad.informatyzacja.recruitmentsystem.service.QuestionService
```
[the remark this answers is not in the document]
Comment (Mario-659) [reply backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/controller/QuestionResource.kt]: I will do that. Do we allow wildcard imports?

[the remark this answers is not in the document]
Comment (Mendrejk) [reply backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/controller/QuestionResource.kt]: Sure, we can use wildcards. Maybe not for recruitmentsystem here as it's just 2 imports but it makes sense for springframework. Actually IntelliJ automatically wildcards imports, but I think 6 or more is the sweetspot when it starts wildcarding. I think it's better to just say import org.springframework.* rather than have a wall of fear everyone will just scroll over and never read

File: backend/recruitment-system/src/test/groovy/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/QuestionServiceSpec.groovy (hunk 1 above)
Comment (Mario-659) [reply backend/recruitment-system/src/test/groovy/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/QuestionServiceSpec.groovy:44]: Aren't tests already self explanatory? I don't really know what to write there. Look at this example
```groovy
    def "should return all questions from repository"() {
        when:
        def actual = questionService.getAllQuestions()

        then:
        1 * questionRepository.findAll() >> questions
        actual == questions
    }
```
I can already read this in plain English (*when questionService getAllQuestions method is called...*), but when I add story:
```groovy
    def "should return all questions from repository"() {
        when: "questionService getAllQuestions method is called"
        def actual = questionService.getAllQuestions()

        then:
        1 * questionRepository.findAll() >> questions
        actual == questions
    }
```
then when I read it it becomes more like this: *when questionService getAllQuestions method is called, questionService getAllQuestions method is called*
The same with then block

Comment (Mendrejk) [reply backend/recruitment-system/src/test/groovy/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/QuestionServiceSpec.groovy:46]: What about something like 
when "All questions are retrieved"
then "The QuestionRepository should be called once and the result should match with the FindAll result"?
But I guess you could argue that it's already self explanatory. If you don't want to then don't write the stories. I just always add stories and never though about whether it's already readable or not.

File: backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/controller/QuestionController.kt [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+import pl.edu.pwr.samorzad.informatyzacja.recruitmentsystem.service.QuestionService
+
+import org.springframework.http.HttpStatus
+import org.springframework.http.ResponseEntity
+import org.springframework.web.bind.annotation.*
+
+import org.slf4j.LoggerFactory
+
+@RestController
+@RequestMapping("/api/v1")
```
Comment (Bravo555) [inline backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/controller/QuestionController.kt:13]: It's fine for now, but we should probably add a task for proper API version routing so it's not repeated for every endpoint

File: backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/exception/TeamIdDoesNotMatchQuestionException.kt [hunk 7]
```diff
@@ -0,0 +1,6 @@
+package pl.edu.pwr.samorzad.informatyzacja.recruitmentsystem.exception
+
+import org.springframework.http.HttpStatus
+import org.springframework.web.server.ResponseStatusException
+
+class TeamIdDoesNotMatchQuestionException : ResponseStatusException(HttpStatus.CONFLICT, "Team id does not match the question id")
```
Comment (Bravo555) [inline backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/exception/TeamIdDoesNotMatchQuestionException.kt:6]: I think it makes more sense to return `404 NOT FOUND` here. This exception is used in `getQuestion` method, for which the endpoint is `teams/{teamId}/questions/{questionId}`. For a person using the endpoint, they don't care and they shouldn't be allowed to know whether the question exists in another team, and more importantly, it can be used to leak how many questions are in the database.

**Review by Bravo555 — CHANGES_REQUESTED** [review]

File: backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/exception/TeamIdDoesNotMatchQuestionException.kt (hunk 7 above)
Comment (Mendrejk) [reply backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/exception/TeamIdDoesNotMatchQuestionException.kt:6]: +1

File: backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/QuestionService.kt [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+            question.content,
+            question.difficulty,
+            question.minimumMinutes,
+            question.maximumMinutes,
+            Team(teamId)
+        )
+
+        return questionRepository.save(questionToSave)
+    }
```
Comment (Mendrejk) [inline backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/QuestionService.kt:49]: use addQuestion instead (also addQuesiton in repository)

**Review by Mendrejk — CHANGES_REQUESTED** [review]
Fix It!

**Comment by Mario-659** [comment]
- added function in repository to simplify logic in service (@Mendrejk to return Nullable value you can override methods and specify return value instead of using `...OrNull` method but keep in mind that for collections it will return empty collection even if given id does not exist (from my own testing, I couldn't find that in docs 😞 ))
- @Bravo555 changed validation of question and team id, now it looks like this: assertTeamExists -> assertQuestionWithGivenTeamIdAndQuestionExists -> (for PUT and POST) validateQuestionBody
- updated tests but they still don't cover every possibility
- exceptions and controller implemented as in tutorial ([repo](https://github.com/petersommerhoff/thenewboston-spring-boot-with-kotlin/blob/main/src/main/kotlin/tv/codealong/tutorials/springboot/thenewboston/controller/BankController.kt))

To discuss:
1. Now when exception occurs in response body is only message from exception. I think much better would be something like default spring answer:
```
{
 "timestamp": 1500597044204,
 "status": 400,
 "error": "Bad Request",
 "exception": "org.springframework.http.converter.HttpMessageNotReadableException",
 "message": "message from exception",
 "path": "/birds"
}
```
I'm not sure how to do this but I can look into that or we can leave it as it is for now.
2. I validate RequestBody in service (`private fun validateQuestion(question: Question)`) however we can use something like [here](https://springframework.guru/bean-validation-in-spring-boot/)

**Review by Bravo555 — APPROVED** [review]

**Review by Mendrejk — APPROVED** [review]
Ship It!

**Comment by Mendrejk** [comment]
Regarding your suggestions, let's leave this as-is for now. We can create Jira's for them

**Review by DocentSzachista — APPROVED** [review]

**Comment by Mendrejk** [comment]
> * added function in repository to simplify logic in service (@Mendrejk to return Nullable value you can override methods and specify return value instead of using `...OrNull` method but keep in mind that for collections it will return empty collection even if given id does not exist (from my own testing, I couldn't find that in docs disappointed ))
> * @Bravo555 changed validation of question and team id, now it looks like this: assertTeamExists -> assertQuestionWithGivenTeamIdAndQuestionExists -> (for PUT and POST) validateQuestionBody
> * updated tests but they still don't cover every possibility
> * exceptions and controller implemented as in tutorial ([repo](https://github.com/petersommerhoff/thenewboston-spring-boot-with-kotlin/blob/main/src/main/kotlin/tv/codealong/tutorials/springboot/thenewboston/controller/BankController.kt))
> 
> To discuss:
> 
> 1. Now when exception occurs in response body is only message from exception. I think much better would be something like default spring answer:
> 
> ```
> {
>  "timestamp": 1500597044204,
>  "status": 400,
>  "error": "Bad Request",
>  "exception": "org.springframework.http.converter.HttpMessageNotReadableException",
>  "message": "message from exception",
>  "path": "/birds"
> }
> ```
> 
> I'm not sure how to do this but I can look into that or we can leave it as it is for now. 2. I validate RequestBody in service (`private fun validateQuestion(question: Question)`) however we can use something like [here](https://springframework.guru/bean-validation-in-spring-boot/)

Jira [SR-27](https://informatyzacja.atlassian.net/browse/SR-27), [SR-28](https://informatyzacja.atlassian.net/browse/SR-28)
